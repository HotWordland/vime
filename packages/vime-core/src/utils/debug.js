/* eslint-disable no-undef */
/* eslint-disable import/no-mutable-exports,no-console */

const noop = () => {}

export let log = noop
export let warn = noop
export let error = noop

if (
  process.env.NODE_ENV === 'development' &&
  window.console !== undefined
) {
  log = function () { console.log('[Vime]', ...arguments) }
  warn = function () { console.warn('[Vime warn]', ...arguments) }
  error = function () { console.error('[Vime error]', ...arguments) }
}
