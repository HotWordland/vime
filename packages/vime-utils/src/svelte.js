import { noop, run_all, SvelteComponent } from 'svelte/internal'
import { createEventDispatcher, onDestroy } from 'svelte'
import { is_instance_of } from './unit'

export const is_svelte_component = input => Boolean(
  input && is_instance_of(input.prototype, SvelteComponent)
)

export const try_create_svelte_dispatcher = () => {
  try {
    return createEventDispatcher()
  } catch (e) {
    return noop
  }
}

export const try_on_svelte_destroy = cb => {
  try {
    onDestroy(cb)
  } catch (e) {}
}

export const on_svelte_instance_destroy = (instance, cb) => {
  if (!is_svelte_component(instance)) return
  if (!instance || !instance.$$ || !instance.$$.on_destroy) return
  instance.$$.on_destroy.push(cb)
}
