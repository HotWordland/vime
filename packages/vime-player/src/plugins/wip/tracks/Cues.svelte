{#if $isCaptionsEnabled}
  <div
    class:active={isCueActive && $isCaptionsActive}
    class:withControls={$isControlsActive}
    class:fullscreen={$isFullscreenActive}
    class:mobile={$isMobile}
    data-testid="cues"
  >
    {#if isCueActive}
      <span data-testid="cue">
        {@html getCueContent(currentCue)}
      </span>
    {/if}
  </div>
{/if}

<script>
  import { getContext, createEventDispatcher } from 'svelte'
  import { element, listen, append } from 'svelte/internal'
  import { warn } from '~utils/debug'
  import { is_number, is_instance_of } from '~utils/unit'
  import { ctxKey } from '~src/context'
  import { isMobile } from '~src/store'
  import PlayerEvent from '~src/PlayerEvent'

  const dispatch = createEventDispatcher()

  const ctx = getContext(ctxKey)
  const isControlsActive = ctx.isControlsActive
  const isCaptionsEnabled = ctx.isCaptionsEnabled
  const isCaptionsActive = ctx.isCaptionsActive
  const isFullscreenActive = ctx.isFullscreenActive
  
  // $: if (!$isAudio && isPlaybackReady && _track) $isCaptionsEnabled = _tracks.length > 0

  const cues = []
  let currentCueIndex = -1
  let currentCue = null
  let isCueActive = false

  export let track
  export let currentTime
  export let crossOrigin

  const shouldCueBeActive = cue => (currentTime >= cue.startTime) && (currentTime <= cue.endTime)

  const getCueContent = cue => {
    const div = element('div')
    append(div, cue.getCueAsHTML())
    return div.innerHTML.trim()
  }

  const findNextCueIndex = start => {
    let index = start
    while (
      index >= 0 &&
      index < (cues.length - 1) &&
      currentTime > cues[index].startTime &&
      !shouldCueBeActive(cues[index])
    ) {
      index += 1
    }
    return index
  }

  export const toList = () => cues

  export const getCurrentCue = () => ({
    index: currentCueIndex,
    cue: currentCue,
    isActive: !!currentCue && shouldCueBeActive(currentCue),
    content: currentCue ? getCueContent(currentCue) : null
  })

  const validateCue = cue => {
    if (!is_instance_of(cue, window.VTTCue)) {
      warn(
        'Cues :: invalid cue must be an instance of window.VTTCue.' +
        '\nsee https://developer.mozilla.org/en-US/docs/Web/API/VTTCue'
      )
      return false
    }
    if (cues.some(c => (c.startTime <= cue.endTime) && (cue.startTime <= c.endTime))) {
      warn('Cues :: invalid cue found, overlapping cues are not supported at this time')
      return false
    }
    return true
  }

  export const addCue = cue => {
    if (!validateCue(cue)) return
    let index = 0
    while (cues[index] && (cues[index].endTime < cue.startTime)) index++
    cues.splice(index, 0, cue)
    onCuesChange()
  }

  export const addCues = cues => { cues.map(addCue) }

  export const removeCue = cue => {
    const index = is_number(cue) ? cue : cues.findIndex(c => c === cue)
    if (index <= 0 || index >= cues.length) {
      warn('Cues :: could not remove cue because it could not be found or index was out of bounds')
      return
    }
    cues.splice(index, 1)
    onCuesChange()
  }

  export const removeAllCues = () => {
    cues.splice(0, cues.length)
    currentCueIndex = -1
    currentCue = null
    onCuesChange()
  }

  const onCueEnter = () => dispatch(PlayerEvent.CUE_ENTER, getCurrentCue())

  const onCueExit = () => {
    if (!isCueActive) return
    isCueActive = false
    dispatch(PlayerEvent.CUE_EXIT, getCurrentCue())
  }

  const onCueChange = index => {
    onCueExit()
    if (currentCueIndex === index) return
    currentCueIndex = index
    currentCue = cues[index]
    dispatch(PlayerEvent.CUE_CHANGE, getCurrentCue())
  }

  const onCuesChange = () => {
    dispatch(PlayerEvent.CUES_CHANGE, [...cues])
    if (currentCueIndex >= 0 && cues[currentCueIndex]) {
      currentCue = cues[currentCueIndex]
      return
    }
    if (cues.length > 0) onCueChange(findNextCueIndex(0))
  }

  // TODO: Fix IE captions if CORS is used.
  // Fetch captions and inject as blobs instead (data URIs not supported).
  const loadCues = () => {
    const media = element('audio')
    media.crossOrigin = crossOrigin
    const trackEl = element('track')
    trackEl.default = true
    trackEl.src = track.src
    listen(trackEl, 'load', () => {
      const newCues = Array.from(media.textTracks[0].cues)
      if (newCues.some(c => !validateCue(c))) return
      cues.push(...newCues)
      dispatch(PlayerEvent.CUES_LOADED, cues)
      onCuesChange()
    }, { once: true })
    append(media, trackEl)
  }

  const onTrackChange = () => {
    cues.splice(0, cues.length)
    currentCueIndex = -1
    currentCue = null
    track ? loadCues() : onCuesChange()
  }

  $: onTrackChange(track)

  $: if (currentCue && (currentCueIndex < cues.length - 1) && currentTime > currentCue.endTime) {
    onCueChange(findNextCueIndex(currentCueIndex))
  }

  $: if (currentCueIndex > 0 && currentTime < cues[currentCueIndex - 1].endTime) {
    onCueChange(findNextCueIndex(0))
  }

  $: isCueActive = (currentCue && currentTime >= 0) ? shouldCueBeActive(currentCue) : false
  $: if (isCueActive) onCueEnter()
</script>

<style type="text/scss">
  @import '../../style/common';

  // Container (cues)
  div {
    bottom: 0;
    color: #fff;
    font-size: $font-size-medium;
    left: 0;
    opacity: 0;
    padding: $control-spacing;
    position: absolute;
    text-align: center;
    transition: transform 0.4s ease-in-out, opacity 0.3s ease;
    width: 100%;

    &.active {
      opacity: 1;
    }

    &.withControls {
      transform: translateY(-($control-spacing * 4) - 8);

      &.mobile {
        transform: translateY(-($control-spacing * 4));
      }
    }

    @media (min-width: $bp-sm) {
      font-size: $font-size-small;
      padding: ($control-spacing * 2);
    }

    @media (min-width: $bp-md) {
      font-size: $font-size-large;
    }

    &.fullscreen {
      @media (min-width: $bp-lg) {
        font-size: $font-size-extra-large;
      }
    }
  }

  // Cue
  span {
    background: $color-gray-600;
    border-radius: 2px;
    box-decoration-break: clone;
    line-height: 185%;
    padding: 0.2em 0.5em;
    white-space: pre-wrap;

    // Firefox adds a <div> when using getCueAsHTML()
    :global(div) {
      display: inline;
    }

    &:empty {
      display: none;
    }
  }
</style>