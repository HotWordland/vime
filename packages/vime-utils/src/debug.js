/* eslint-disable import/no-mutable-exports,no-console */

import { noop } from 'svelte/internal';

export let log = noop;
export let warn = noop;
export let error = noop;

if (
  // eslint-disable-next-line no-undef
  process.env.NODE_ENV === 'development' &&
  window.console !== undefined
) {
  log = function () { console.log('[Vime]', ...arguments); };
  warn = function () { console.warn('[Vime warn]', ...arguments); };
  error = function () { console.error('[Vime error]', ...arguments); };
}
