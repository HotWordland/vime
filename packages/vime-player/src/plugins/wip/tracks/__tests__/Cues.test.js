import { act } from '@testing-library/svelte'
import { warn } from '~utils/debug'
import Cues from '../Cues.svelte'

jest.mock('~utils/debug.js')

class Cue {
  constructor (startTime, endTime, text) {
    this.startTime = startTime
    this.endTime = endTime
    this.text = text
  }

  getCueAsHTML () {
    const span = document.createElement('span')
    span.textContent = this.text
    return span
  }
}

window.VTTCue = Cue

describe('components', () => {
  describe('Cues', () => {
    let cues

    const Event = Object.freeze({
      CUE_CHANGE: 'cuechange',
      CUES_LOADED: 'cuesloaded',
      CUES_CHANGE: 'cueschange',
      CUE_ENTER: 'cueenter',
      CUE_EXIT: 'cueexit'
    })

    const render = props => {
      // const results = renderWithContext({
      //   track: null,
      //   currentTime: 0,
      //   Component: Cues,
      //   ...props
      // })
      const results = {}
      listen(results.component, Object.values(Event))
      return results
    }

    beforeEach(() => {
      cues = [
        new Cue(3, 5, '#1 3-5'),
        new Cue(6, 10, '#2 6-10'),
        new Cue(15, 17, '#3 15-17'),
        new Cue(40, 43, '#4 40-43')
      ]
    })

    it('should load cues', () => {
      jest.spyOn(window.HTMLMediaElement.prototype, 'textTracks', 'get').mockReturnValue([{ cues }])
      const trackAddEventSpy = jest.spyOn(window.HTMLTrackElement.prototype, 'addEventListener')
      const { component } = render({ track: { src: '/src' } })

      trackAddEventSpy.mock.calls.pop()[1]() // Fire load event on HTMLTrackElement

      expect(component.toList()).toEqual(cues)
      expect(component).toHaveFiredEventWith(Event.CUES_LOADED, cues)
      expect(component).toHaveFiredEventTimes(Event.CUES_LOADED, 1)
      expect(component).toHaveFiredEventWith(Event.CUES_CHANGE, cues)
      expect(component).toHaveFiredEventWith(Event.CUE_CHANGE, {
        index: 0,
        cue: cues[0],
        isActive: false,
        content: expect.stringContaining(cues[0].text)
      })
      expect(component).not.toHaveFiredEvent(Event.CUE_ENTER)
      expect(component).not.toHaveFiredEvent(Event.CUE_EXIT)
    })

    it('should add a cue', () => {
      const { component } = render()
      const list = component.toList()

      component.addCue(cues[0])

      const expectedCurrentCue = {
        index: 0,
        cue: cues[0],
        isActive: false,
        content: expect.stringContaining(cues[0].text)
      }

      expect(list).toEqual([cues[0]])
      expect(list).toHaveLength(1)
      expect(component.getCurrentCue()).toEqual(expectedCurrentCue)
      expect(component).toHaveFiredEventWith(Event.CUES_CHANGE, [cues[0]])
      expect(component).toHaveFiredEventTimes(Event.CUES_CHANGE, 1)
      expect(component).toHaveFiredEventWith(Event.CUE_CHANGE, expectedCurrentCue)
    })

    it('should fail adding a cue if invalid', () => {
      const { component } = render()
      const list = component.toList()

      component.addCue({})

      expect(warn).toHaveBeenCalledWith(expect.stringMatching(/Cues :: invalid cue/))
      expect(list).toEqual([])
      expect(list).toHaveLength(0)
      expect(component).not.toHaveFiredEvent(Event.CUES_CHANGE)
      expect(component).not.toHaveFiredEvent(Event.CUE_CHANGE)
    })

    it('should add all cues', () => {
      const { component } = render()
      const list = component.toList()

      component.addCues(cues)

      expect(list).toEqual(cues)
      expect(list).toHaveLength(cues.length)
      expect(component).toHaveFiredEventWith(Event.CUES_CHANGE, cues)
      expect(component).toHaveFiredEventTimes(Event.CUES_CHANGE, cues.length)
      expect(component).toHaveFiredEventWith(Event.CUE_CHANGE, {
        index: 0,
        cue: cues[0],
        isActive: false,
        content: expect.stringContaining(cues[0].text)
      })
      expect(component).not.toHaveFiredEvent(Event.CUE_ENTER)
      expect(component).not.toHaveFiredEvent(Event.CUE_EXIT)
    })

    it('should add cue in correct position', () => {
      const { component } = render()
      const list = component.toList()

      component.addCues(cues)

      // start
      const cueA = new Cue(0, 2, '#A 0-2')
      component.addCue(cueA)
      expect(list).toEqual([cueA, cues[0], cues[1], cues[2], cues[3]])

      // middle
      const cueB = new Cue(11, 13, '#B 11-13')
      component.addCue(cueB)
      expect(list).toEqual([cueA, cues[0], cues[1], cueB, cues[2], cues[3]])

      // end
      const cueC = new Cue(50, 55, '#C 50-55')
      component.addCue(cueC)
      expect(list).toEqual([cueA, cues[0], cues[1], cueB, cues[2], cues[3], cueC])
    })

    it('should remove a cue', () => {
      const { component } = render()
      const list = component.toList()

      component.addCues(cues)
      component.removeCue(cues[0])

      const newCues = [cues[1], cues[2], cues[3]]
      expect(list).toEqual(newCues)
      expect(list).toHaveLength(newCues.length)
      expect(component).toHaveFiredEventWith(Event.CUES_CHANGE, cues)
      expect(component).toHaveFiredEventWith(Event.CUES_CHANGE, newCues)
      expect(component).toHaveFiredEventTimes(Event.CUES_CHANGE, cues.length + 1)
    })

    it('should remove all cues', () => {
      const { component } = render()
      const list = component.toList()

      component.addCues(cues)
      component.removeAllCues()

      expect(list).toEqual([])
      expect(list).toHaveLength(0)
      expect(component.getCurrentCue()).toEqual({
        index: -1,
        cue: null,
        isActive: false,
        content: null
      })
      expect(component).toHaveFiredEventWith(Event.CUES_CHANGE, cues)
      expect(component).toHaveFiredEventWith(Event.CUES_CHANGE, [])
      expect(component).toHaveFiredEventTimes(Event.CUES_CHANGE, cues.length + 1)
    })

    it('should reset if track is set to null', async () => {
      const { component } = render()
      const list = component.toList()

      component.track = ''
      component.addCues(cues)
      component.track = null

      expect(list).toEqual([])
      expect(list).toHaveLength(0)
      expect(component.getCurrentCue()).toEqual({
        index: -1,
        cue: null,
        isActive: false,
        content: null
      })
      expect(component).toHaveFiredEventWith(Event.CUES_CHANGE, [])
    })

    it('should keep reference to the same array when calling toList', () => {
      const { component } = render()
      const list = component.toList()

      component.addCues(cues)
      expect(list).toEqual(cues)

      component.removeCue(cues[1])
      expect(list).toEqual([cues[0], cues[2], cues[3]])

      component.removeAllCues()
      expect(list).toEqual([])
    })

    it('should update active cue when current time moves forward', async () => {
      let currentCue

      const { component, container } = render()
      component.addCues(cues)

      // Set cue 0
      await act(() => { component.currentTime = 4 })
      currentCue = component.getCurrentCue()
      expect(currentCue).toEqual({
        index: 0,
        cue: cues[0],
        isActive: true,
        content: expect.anything()
      })
      expect(component).toHaveFiredEventWith(Event.CUE_ENTER, currentCue)
      expect(component).not.toHaveFiredEvent(Event.CUE_EXIT)

      // Set cue 1
      await act(() => { component.currentTime = 6 })
      expect(component).toHaveFiredEventWith(Event.CUE_EXIT, {
        ...currentCue,
        isActive: false
      })
      currentCue = component.getCurrentCue()
      expect(currentCue).toEqual({
        index: 1,
        cue: cues[1],
        isActive: true,
        content: expect.anything()
      })
      expect(component).toHaveFiredEventWith(Event.CUE_ENTER, currentCue)

      // Set cue 3
      await act(() => { component.currentTime = 41 })
      expect(component).toHaveFiredEventWith(Event.CUE_EXIT, {
        ...currentCue,
        isActive: false
      })
      currentCue = component.getCurrentCue()
      expect(currentCue).toEqual({
        index: 3,
        cue: cues[3],
        isActive: true,
        content: expect.anything()
      })
      expect(component).toHaveFiredEventWith(Event.CUE_ENTER, currentCue)
    })

    it('should update active cue when current time moves backward', async () => {
      let currentCue

      const { component, container } = render()
      component.addCues(cues)

      // Set cue 3
      await act(() => { component.currentTime = 41 })
      currentCue = component.getCurrentCue()
      expect(component.getCurrentCue()).toEqual({
        index: 3,
        cue: cues[3],
        isActive: true,
        content: expect.anything()
      })
      expect(component).toHaveFiredEventWith(Event.CUE_ENTER, currentCue)

      // Set cue 1
      await act(() => { component.currentTime = 8 })
      expect(component).toHaveFiredEventWith(Event.CUE_EXIT, {
        ...currentCue,
        isActive: false
      })
      currentCue = component.getCurrentCue()
      expect(currentCue).toEqual({
        index: 1,
        cue: cues[1],
        isActive: true,
        content: expect.anything()
      })
      expect(component).toHaveFiredEventWith(Event.CUE_ENTER, currentCue)
    })

    it('should look structurally correct when there is no active cue', async () => {
      const { container, context } = render()
      await act(() => { context.isCaptionsEnabled.set(true) })
      expect(container.firstChild).toMatchSnapshot()
    })

    it('should look structurally correct when there is a active cue', async () => {
      const { container, component, context } = render()
      component.addCues(cues)

      await act(() => {
        context.isCaptionsEnabled.set(true)
        component.currentTime = cues[0].startTime
      })

      expect(container.firstChild).toMatchSnapshot()
    })
  })
})
