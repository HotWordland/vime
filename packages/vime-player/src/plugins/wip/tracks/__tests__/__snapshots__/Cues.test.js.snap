// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components Cues should look structurally correct when there is a active cue 1`] = `
<div>
  <div
    class="svelte-v278vs active notMobile"
    data-testid="cues"
  >
    <span
      class="svelte-v278vs"
      data-testid="cue"
    >
      <span>
        #1 3-5
      </span>
    </span>
  </div>
  
  
</div>
`;

exports[`components Cues should look structurally correct when there is no active cue 1`] = `
<div>
  <div
    class="svelte-v278vs notMobile"
    data-testid="cues"
  />
  
  
</div>
`;
